src/wallet/walletdb.cpp

```diff
@@ -540,30 +540,45 @@ static LoadResult LoadRecords(CWallet* pwallet, DatabaseBatch& batch, const std:
     return LoadRecords(pwallet, batch, key, prefix, load_func);
 }
 
+bool HasLegacyRecords(CWallet* pwallet)
+{
+    const auto& batch = pwallet->GetDatabase().MakeBatch();
+    return HasLegacyRecords(pwallet, *batch);
+}
+
+bool HasLegacyRecords(CWallet* pwallet, DatabaseBatch& batch)
+{
+    for (const auto& type : DBKeys::LEGACY_TYPES) {
+        DataStream key;
+        DataStream value{};
+        DataStream prefix;
+
+        prefix << type;
+        std::unique_ptr<DatabaseCursor> cursor = batch.GetNewPrefixCursor(prefix);
+        if (!cursor) {
+            // Could only happen on a closed db, which means there is an error in the code flow.
+            pwallet->WalletLogPrintf("Error getting database cursor for '%s' records", type);
+            throw std::runtime_error(strprintf("Error getting database cursor for '%s' records", type));
+        }
+
+        DatabaseCursor::Status status = cursor->Next(key, value);
+        if (status != DatabaseCursor::Status::DONE) {
```

Comment:
nit, maybe this should be: 

```suggestion
        if (status == DatabaseCursor::Status::MORE)
```

as is done in [`WalletBatch::IsEncrypted()`](https://github.com/bitcoin/bitcoin/blob/1a6fc04d815e45ee895bd7ff0bc1b03685ed96c3/src/wallet/walletdb.cpp#L193-L202)